x/ibc/06-solomachine/types/client_state.go

```diff
@@ -0,0 +1,438 @@
+	channelexported "github.com/cosmos/cosmos-sdk/x/ibc/04-channel/exported"
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/04-channel/types"
+	commitmentexported "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment/exported"
+	commitmenttypes "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment/types"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/24-host"
+)
+
+var _ clientexported.ClientState = ClientState{}
+
+// ClientState of a Solo Machine represents whether or not the client is frozen.
+type ClientState struct {
+	// Client ID
+	ID string `json:"id" yaml:"id"`
+
+	// Frozen status of the client
+	Frozen bool `json:"frozen" yaml:"frozen"`
+
+	// Current consensus state of the client
+	ConsensusState ConsensusState `json:"consensus_state" yaml:"consensus_state"`
+}
+
+// InitializeFromMsg creates a solo machine client from a MsgCreateClient
+func InitializeFromMsg(msg MsgCreateClient) (ClientState, error) {
+	return Initialize(msg.GetClientID(), msg.ConsensusState)
+}
+
+// Initialize creates an unfrozen client with the initial consensus state
+func Initialize(id string, consensusState ConsensusState) (ClientState, error) {
+	return ClientState{
+		ID:             id,
+		Frozen:         false,
+		ConsensusState: consensusState,
+	}, nil
+}
+
+// NewClientState creates a new ClientState instance.
+func NewClientState(id string, consensusState ConsensusState) ClientState {
+	return ClientState{
+		ID:             id,
+		Frozen:         false,
+		ConsensusState: consensusState,
+	}
+}
+
+// GetID returns the solo machine client state identifier.
+func (cs ClientState) GetID() string {
+	return cs.ID
+}
+
+// GetChainID returns an empty string.
+func (cs ClientState) GetChainID() string {
+	return ""
+}
+
+// ClientType is Solo Machine.
+func (cs ClientState) ClientType() clientexported.ClientType {
+	return clientexported.SoloMachine
+}
+
+// GetLatestHeight returns the latest sequence number.
+func (cs ClientState) GetLatestHeight() uint64 {
+	return cs.ConsensusState.Sequence
+}
+
+// IsFrozen returns true if the client is frozen
+func (cs ClientState) IsFrozen() bool {
+	return cs.Frozen
+}
+
+// Validate performs basic validation of the client state fields.
+func (cs ClientState) Validate() error {
+	if err := host.ClientIdentifierValidator(cs.ID); err != nil {
+		return err
+	}
+	return cs.ConsensusState.ValidateBasic()
+}
+
+// VerifyClientConsensusState verifies a proof of the consensus state of the
+// Solo Machine client stored on the target machine.
+func (cs ClientState) VerifyClientConsensusState(
+	store sdk.KVStore,
+	cdc *codec.Codec,
+	root commitmentexported.Root,
+	_ uint64,
+	counterpartyClientIdentifier string,
+	consensusHeight uint64,
+	prefix commitmentexported.Prefix,
+	proof commitmentexported.Proof,
+	consensusState clientexported.ConsensusState,
+) error {
+	clientPrefixedPath := "clients/" + counterpartyClientIdentifier + "/" + host.ConsensusStatePath(consensusHeight)
```

Comment:
Can we abstract out some of this repeated logic (proof casting, checking whether the client is frozen, etc.) to a common helper function?